### PR TITLE
Add functional tests for account PriceGranularity config

### PR DESCRIPTION
**Pull request by osulzhenko**
### 🔧 Type of changes
- [ ] new bid adapter
- [ ] update bid adapter
- [ ] new feature
- [ ] new analytics adapter
- [ ] new module
- [ ] bugfix
- [ ] documentation
- [ ] configuration
- [x] tech debt (test coverage, refactorings, etc.)

### ✨ What's the context?

What's the context for the changes? Are there any

### 🧠 Rationale behind the change

Why did you choose to make these changes? Were there any trade-offs you had to consider?

### 🧪 Test plan

How do you know the changes are safe to ship to production?

### 🏎 Quality check

- [ ] Are your changes following [our code style guidelines](https://github.com/prebid/prebid-server-java/blob/master/docs/developers/code-style.md)?
- [ ] Are there any breaking changes in your code?
- [ ] Does your test coverage exceed 90%?
- [ ] Are there any erroneous console logs, debuggers or leftover code in your changes?

### CODE REVIEW

File: src/test/groovy/org/prebid/server/functional/model/config/PriceGranularityType.groovy [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+    HIGH(2, [Range.getDefault(20, 0.01)]),
+    AUTO(2, [Range.getDefault(5, 0.05), Range.getDefault(10, 0.1), Range.getDefault(20, 0.5)]),
+    DENSE(2, [Range.getDefault(3, 0.01), Range.getDefault(8, 0.05), Range.getDefault(20, 0.5)]),
+    UNKNOWN(null, [])
+
+    final Integer precision
+    final List<Range> ranges
+
+    PriceGranularityType(Integer precision, List<Range> ranges) {
+        this.precision = precision
```
Comment (marki1an) [inline src/test/groovy/org/prebid/server/functional/model/config/PriceGranularityType.groovy:20]: Do we need here precision when it always "2"?

File: src/test/groovy/org/prebid/server/functional/tests/TargetingSpec.groovy [hunk 2]
```diff
@@ -1121,6 +1126,204 @@ class TargetingSpec extends BaseSpec {
...
+        then: "Request should fail with an error"
+        def exception = thrown(PrebidServerException)
+        assert exception.statusCode == BAD_REQUEST.code()
+        assert exception.responseBody == 'Invalid request format: Price granularity error: empty granularity definition supplied'
+    }
+
+    def "PBS should prioritize price granularity from original request over account config"() {
+        given: "Default bidRequest with price granularity"
+        def bidRequest = BidRequest.defaultBidRequest.tap {
+            ext.prebid.targeting = new Targeting(priceGranularity: PriceGranularity.getDefault(priceGranularity))
```
Comment (marki1an) [inline src/test/groovy/org/prebid/server/functional/tests/TargetingSpec.groovy:1153]: I suppose we always use "var" for object that gonna be in the assert block

File: src/test/groovy/org/prebid/server/functional/tests/TargetingSpec.groovy [hunk 3]
```diff
@@ -1121,6 +1126,204 @@ class TargetingSpec extends BaseSpec {
...
+        when: "PBS processes auction request"
+        defaultPbsService.sendAuctionRequest(bidRequest)
+
+        then: "Request should fail with an error"
+        def exception = thrown(PrebidServerException)
+        assert exception.statusCode == BAD_REQUEST.code()
+        assert exception.responseBody == 'Invalid request format: Price granularity error: empty granularity definition supplied'
+    }
+
+    def "PBS should prioritize price granularity from original request over account config"() {
```
Comment (marki1an) [inline src/test/groovy/org/prebid/server/functional/tests/TargetingSpec.groovy:1150]: `PBS auction should prioritize price granularity from original request over account config`

File: src/test/groovy/org/prebid/server/functional/tests/TargetingSpec.groovy [hunk 4]
```diff
@@ -1121,6 +1126,204 @@ class TargetingSpec extends BaseSpec {
         assert targeting["hb_env"] == HB_ENV_AMP
     }
 
+    def "PBS should throw error when price granularity from original request is empty"() {
```
Comment (marki1an) [inline src/test/groovy/org/prebid/server/functional/tests/TargetingSpec.groovy:1129]: `PBS auction should throw error when price granularity from original request is empty`
In addition you also can add same test for amp endpoint

File: src/test/groovy/org/prebid/server/functional/tests/TargetingSpec.groovy [hunk 5]
```diff
@@ -1121,6 +1126,204 @@ class TargetingSpec extends BaseSpec {
...
+
+        then: "BidderRequest should include price granularity from bidRequest"
+        def bidderRequest = bidder.getBidderRequest(ampStoredRequest.id)
+        assert bidderRequest?.ext?.prebid?.targeting?.priceGranularity == ampStoredRequest?.ext?.prebid?.targeting?.priceGranularity
+
+        where:
+        priceGranularity << PriceGranularityType.values() - UNKNOWN
+    }
+
+    def "PBS should include price granularity from account config when original request doesn't contain it"() {
```
Comment (marki1an) [inline src/test/groovy/org/prebid/server/functional/tests/TargetingSpec.groovy:1204]: `"PBS auction should include price granularity from account config when original request doesn't contain price granularity"`

File: src/test/groovy/org/prebid/server/functional/tests/TargetingSpec.groovy [hunk 6]
```diff
@@ -1121,6 +1126,204 @@ class TargetingSpec extends BaseSpec {
...
+        given: "Default bidRequest with empty price granularity"
+        def bidRequest = BidRequest.defaultBidRequest.tap {
+            ext.prebid.targeting = new Targeting(priceGranularity: PriceGranularity.getDefault(UNKNOWN))
+        }
+
+        and: "Account in the DB"
+        def accountAuctionConfig = new AccountAuctionConfig(priceGranularity: PBSUtils.getRandomEnum(PriceGranularityType))
+        def accountConfig = new AccountConfig(status: ACTIVE, auction: accountAuctionConfig)
+        def account = new Account(uuid: bidRequest.accountId, config: accountConfig)
+        accountDao.save(account)
```
Comment (marki1an) [inline src/test/groovy/org/prebid/server/functional/tests/TargetingSpec.groovy:1139]: Please create a separate method for creating an account with priceGranularity

File: src/test/groovy/org/prebid/server/functional/tests/TargetingSpec.groovy [hunk 7]
```diff
@@ -1121,6 +1126,204 @@ class TargetingSpec extends BaseSpec {
...
+
+        then: "BidderRequest should include price granularity from account config"
+        def bidderRequest = bidder.getBidderRequest(bidRequest.id)
+        assert bidderRequest?.ext?.prebid?.targeting?.priceGranularity == PriceGranularity.getDefault(priceGranularity)
+
+        where:
+        priceGranularity << PriceGranularityType.values() - UNKNOWN
+    }
+
+    def "PBS should include price granularity from account config with different name case when original request doesn't contain it"() {
```
Comment (marki1an) [inline src/test/groovy/org/prebid/server/functional/tests/TargetingSpec.groovy:1227]: "PBS auction should include price granularity from account config with different name case when original request doesn't contain price granularity"

**Review by marki1an — CHANGES_REQUESTED** [review]

File: src/test/groovy/org/prebid/server/functional/model/config/PriceGranularityType.groovy (hunk 1 above)
Comment (osulzhenko) [reply src/test/groovy/org/prebid/server/functional/model/config/PriceGranularityType.groovy:20]: some of the tests use null precision

**Review by marki1an — APPROVED** [review]

**Pull request merged**